nanis/ml_exercises/scripts/week_3.py

```diff
@@ -0,0 +1,142 @@
+                    options={'maxiter':400})
+
+def accuracy(theta, x, y, threshold): 
+    probability = sigmoid(x @ theta.T) >= threshold
+    probability = probability.astype('int') 
+    return 100 * sum(probability == y.ravel())/probability.size
+
+def plot_data(data, xlabel, ylabel): 
+    def get_x(data): 
+        return data.iloc[:, 0]
+
+    def get_y(data): 
+        return data.iloc[:, 1]
+
+    admitted = data[data.iloc[:, -1] == 1]
+    not_admitted = data[data.iloc[:, -1] == 0]
+    plt.scatter(get_x(admitted), get_y(admitted), marker='+', color='black', label='passed', edgecolor='black')
+    plt.scatter(get_x(not_admitted), get_y(not_admitted), marker='o', color='yellow', label='failed admitted') 
+    plt.legend(loc='upper right')
+    plt.xlabel(xlabel)
+    plt.ylabel(ylabel)
+    plt.title('figure 1: scatter plot of training data')
+
+    return plt.show() 
+
+def insert_polynomial_features(x): 
+    x = x[:, 1:3]
+    polynomial_features = PolynomialFeatures(6)
+    return polynomial_features.fit_transform(x)
+
+def compute_regularised_cost(theta, lambda_value, x, y):
+    number_of_samples = y.size
+    predictions = sigmoid(x @ theta)
+
+    negative_cost = np.log(predictions).T @ (y)
+    positive_cost = np.log(1-predictions).T @ (1-y)
+    regularisation = (lambda_value/(2*m))*np.sum(np.square(theta[1:]))
+
+    cost = (-1/number_of_samples) * (negative_cost + positive_cost) + regularisation 
+    return sum(cost)
+
+def compute_regularised_gradient(theta, lambda_value, x, y):
+    number_of_samples = y.size
+    predictions = sigmoid(x @ theta.reshape(-1,1))
+
+    intial_gradient = (1/number_of_samples) * x.T @ (predictions-y) 
+    regularisation = (lambda_value/number_of_samples) * np.r_[[[0]], theta[1:].reshape(-1,1)]
+    gradient = initial_gradient + regularisation
+    return gradient.flatten()
+
+
+if __name__ == '__main__':
+    # PART 1
+    file_path = 'data/ex2data1.txt'
+    variables = ['score_1', 'score_2', 'admitted']
+    exam_data = clean_data(file_path, variables)
+
+    xlabel = ('exam_score_1')
+    ylabel = ('exam_score_2')
+    plot_data(exam_data, xlabel, ylabel)
+
+    x, y = extract_features(exam_data)
+    number_of_samples, features = x.shape
+    theta = np.zeros(features)
+
+    cost = compute_cost(theta, x, y)
+    optimised_cost_function = optimised_cost(theta, x, y)
+
+    threshold = 0.5
+    exam_score_samples = np.array([1, 45, 85])
+    optimised_theta = optimised_cost_function.x
+    sigmoid(np.array(exam_score_samples).dot(optimised_theta.T))
+    accuracy = accuracy(optimised_theta, x, y, threshold)
+
+    # plot decision boundary 
+
+
+    # PART 2 
+    file_path = 'data/ex2data2.txt' 
+    variables = ['test_2', 'test_2', 'pass']
+    microchip_data = clean_data(file_path, variables)
+
+    xlabel = ('microchip test 1')
+    ylabel = ('microchip test 2')
+    plot_data(microchip_data, xlabel, ylabel)
+
+    x, y = extract_features(microchip_data)
```

Comment:
Avoid single-letter variable names especially when the scope is more than one line.